File: internal/integrations/fake/fake.go
// Package fake is a test integration. It talks to nothing. It exists so the
// engine, server and command can be exercised end to end, and so a fresh
// deployment can be smoke-tested before any real connection is added.
package fake

import (
	"context"
	"errors"
	"strings"

	"github.com/roee-hersh/hallpass/internal/catalog"
	"github.com/roee-hersh/hallpass/internal/integration"
)

// Integration is the fake product.
type Integration struct{}

// Name is "fake".
func (Integration) Name() string { return "fake" }

// Fields: users, admins, fail.
func (Integration) Fields() []integration.Field {
	return []integration.Field{
		{Name: "users", Description: "comma-separated emails that exist and may read"},
		{Name: "admins", Description: "comma-separated emails allowed every action"},
		{Name: "fail", Default: "none", Enum: []string{"none", "upstream_timeout", "upstream_error", "credential_rejected", "upstream_rate_limited"},
			Description: "make every call fail with this code (for testing callers)"},
	}
}

// Actions of the fake integration.
func (Integration) Actions() []catalog.Action {
	return []catalog.Action{
		{Name: "thing.read", Description: "read a thing (any known user)"},
		{Name: "thing.write", Description: "write a thing (admins only)"},
		{Name: "thing.admin", Description: "administer a thing (admins only)"},
	}
}

// New builds a connection.
func (Integration) New(_ context.Context, s *integration.Settings, _ integration.Deps) (integration.Connection, error) {
	c := &Connection{users: set(s.Get("users")), admins: set(s.Get("admins"))}
	if f := s.Get("fail"); f != "none" && f != "" {
		c.fail = integration.Code(f)
	}
	return c, nil
}

func set(csv string) map[string]bool {
	m := map[string]bool{}
	for _, p := range strings.Split(csv, ",") {
		p = strings.ToLower(strings.TrimSpace(p))
		if p != "" {
			m[p] = true
		}
	}
	return m
}

// Connection is one fake system.
type Connection struct {
	users, admins map[string]bool
	fail          integration.Code
}

// ResolveIdentity: known users and admins resolve; an email containing
// "ambiguous" is ambiguous; everything else has no account.
func (c *Connection) ResolveIdentity(_ context.Context, u integration.User) (integration.Identity, error) {
	if c.fail != "" {
		return integration.Identity{}, integration.Errorf(c.fail, "fake failure")
	}
	email := strings.ToLower(u.Email)
	if strings.Contains(email, "ambiguous") {
		return integration.Identity{}, integration.UserAmbiguous("several accounts match %s", u.Email)
	}
	if !c.users[email] && !c.admins[email] {
		return integration.Identity{}, integration.UserNotFound("no account for %s", u.Email)
	}
	role := "user"
	if c.admins[email] {
		role = "admin"
	}
	return integration.Identity{ID: email, Display: email, Attrs: map[string]string{"role": role}}, nil
}

// Check evaluates the three actions on "thing:<id>" resources.
// The id "hidden" is not visible; "broken" is unsupported.
func (c *Connection) Check(_ context.Context, r integration.CheckRequest) (integration.Decision, error) {
	if c.fail != "" {
		return integration.Decision{}, integration.Errorf(c.fail, "fake failure")
	}
	if r.Resource.Type != "thing" || r.Resource.ID == "" {
		return integration.Decision{}, integration.Errorf(integration.CodeInvalidRequest, "resource must be thing:<id>")
	}
	switch r.Resource.ID {
	case "hidden":
		return integration.UnknownDecision(integration.CodeResourceNotVisible, "thing %q is not visible to the fake credential", r.Resource.ID), nil
	case "broken":
		return integration.Unsupported("thing %q uses a policy the fake integration cannot evaluate", r.Resource.ID), nil
	}
	admin := r.Identity.Attr("role") == "admin"
	switch r.Action.Name {
	case "thing.read":
		return integration.Allowed("%s is a known user", r.Identity.Display), nil
	case "thing.write", "thing.admin":
		if admin {
			return integration.Allowed("%s is an admin", r.Identity.Display), nil
		}
		return integration.Denied("%s is not an admin", r.Identity.Display), nil
	}
	return integration.Decision{}, errors.New("unreachable: unknown action")
}

// Probe reports the configured user counts.
func (c *Connection) Probe(context.Context) (integration.ProbeResult, error) {
	if c.fail != "" {
		return integration.ProbeResult{}, integration.Errorf(c.fail, "fake failure")
	}
	var warnings []string
	if len(c.users)+len(c.admins) == 0 {
		warnings = append(warnings, "no users or admins configured; every check will answer user_not_found")
	}
	return integration.ProbeResult{
		Summary:  "fake integration ready",
		Warnings: warnings,
	}, nil
}

File: hallpass-py/src/hallpass/integrations/fake.py
"""A test integration. It talks to nothing. It exists so the engine, server
and command can be exercised end to end, and so a fresh deployment can be
smoke-tested before any real connection is added."""

from __future__ import annotations

from hallpass.core.catalog import Action
from hallpass.core.context import Context
from hallpass.core.decision import (
    Code,
    Decision,
    allowed,
    denied,
    errorf,
    unknown_decision,
    unsupported,
    user_ambiguous,
    user_not_found,
)
from hallpass.core.integration import CheckRequest, Connection, Deps, Field, Identity, Integration, ProbeResult, Settings, User


class Fake(Integration):
    def name(self) -> str:
        return "fake"

    def fields(self) -> list[Field]:
        return [
            Field(name="users", description="comma-separated emails that exist and may read"),
            Field(name="admins", description="comma-separated emails allowed every action"),
            Field(
                name="fail",
                default="none",
                enum=("none", "upstream_timeout", "upstream_error", "credential_rejected", "upstream_rate_limited"),
                description="make every call fail with this code (for testing callers)",
            ),
        ]

    def actions(self) -> list[Action]:
        return [
            Action("thing.read", "read a thing (any known user)"),
            Action("thing.write", "write a thing (admins only)"),
            Action("thing.admin", "administer a thing (admins only)"),
        ]

    def new(self, ctx: Context, s: Settings, d: Deps) -> Connection:
        fail = s.get("fail")
        return FakeConnection(_set(s.get("users")), _set(s.get("admins")), Code(fail) if fail not in ("", "none") else None)


def _set(csv: str) -> frozenset[str]:
    return frozenset(p.strip().lower() for p in csv.split(",") if p.strip())


class FakeConnection(Connection):
    def __init__(self, users: frozenset[str], admins: frozenset[str], fail: Code | None) -> None:
        self.users = users
        self.admins = admins
        self.fail = fail

    def resolve_identity(self, ctx: Context, u: User) -> Identity:
        """Known users and admins resolve; an email containing "ambiguous"
        is ambiguous; everything else has no account."""
        if self.fail is not None:
            raise errorf(self.fail, "fake failure")
        email = u.email.lower()
        if "ambiguous" in email:
            raise user_ambiguous(f"several accounts match {u.email}")
        if email not in self.users and email not in self.admins:
            raise user_not_found(f"no account for {u.email}")
        role = "admin" if email in self.admins else "user"
        return Identity(id=email, display=email, attrs={"role": role})

    def check(self, ctx: Context, r: CheckRequest) -> Decision:
        """The three actions on "thing:<id>". The id "hidden" is not
        visible; "broken" is unsupported."""
        if self.fail is not None:
            raise errorf(self.fail, "fake failure")
        if r.resource.type != "thing" or r.resource.id == "":
            raise errorf(Code.INVALID_REQUEST, "resource must be thing:<id>")
        if r.resource.id == "hidden":
            return unknown_decision(Code.RESOURCE_NOT_VISIBLE, f'thing "{r.resource.id}" is not visible to the fake credential')
        if r.resource.id == "broken":
            return unsupported(f'thing "{r.resource.id}" uses a policy the fake integration cannot evaluate')
        admin = r.identity.attr("role") == "admin"
        if r.action.name == "thing.read":
            return allowed(f"{r.identity.display} is a known user")
        if r.action.name in ("thing.write", "thing.admin"):
            if admin:
                return allowed(f"{r.identity.display} is an admin")
            return denied(f"{r.identity.display} is not an admin")
        raise RuntimeError("unreachable: unknown action")

    def probe(self, ctx: Context) -> ProbeResult:
        if self.fail is not None:
            raise errorf(self.fail, "fake failure")
        warnings: tuple[str, ...] = ()
        if not self.users and not self.admins:
            warnings = ("no users or admins configured; every check will answer user_not_found",)
        return ProbeResult(summary="fake integration ready", warnings=warnings)
